src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,113 @@
+    private static final String MISSING_ELEMENT = "The element in the list is missing.";
+    private Object[] values;
+    private int size;
+    private int capacity;
 
+    {
+        values = new Object[DEFAULT_CAPACITY];
+        capacity = DEFAULT_CAPACITY;
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(T value) {
+        if (size == capacity) {
+            increaseCapasity();
+        }
+        values[size] = value;
+        size++;
+    }
 
+    @Override
+    public void add(T value, int index) throws ArrayListIndexOutOfBoundsException {
+        checkIndex(index > size || index < 0);
+        if (index == size) {
+            add(value);
+        } else {
+            Object[] temp = Arrays.copyOfRange(values, index, size);
+            values[index] = value;
+            size = index + 1;
+            addArray(temp);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public T get(int index) throws ArrayListIndexOutOfBoundsException {
+        checkIndex(index >= size || index < 0);
+        return (T) values[index];
     }
 
     @Override
-    public void set(T value, int index) {
-
+    public void set(T value, int index) throws ArrayListIndexOutOfBoundsException {
+        checkIndex(index >= size || index < 0);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index >= size || index < 0);
+        T removedElement = (T) values[index];
+        Object[] temp = Arrays.copyOfRange(values, index + 1, size);
+        size = index;
+        addArray(temp);
+        return removedElement;
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public T remove(T element) throws NoSuchElementException {
+        boolean isExist = false;
+        for (int i = 0; i < size; i++) {
+            if (element == values[i] || element != null && element.equals(values[i])) {
+                remove(i);
+                isExist = true;
+                break;
+            }
+        }
+        if (!isExist) {
+            throw new NoSuchElementException(MISSING_ELEMENT);
+        }
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void increaseCapasity() {
+        capacity *= GROWTH_FACTOR;
+        values = Arrays.copyOf(values, capacity);
+    }
+
+    private void checkIndex(boolean incorrectIndex) throws ArrayListIndexOutOfBoundsException {
+        if (incorrectIndex) {
+            throw new ArrayListIndexOutOfBoundsException(INDEX_ERROR);
+        }
+    }
```

Comment:
move whole logic of index checking and exception throwing in or out of this method
do not split it in half